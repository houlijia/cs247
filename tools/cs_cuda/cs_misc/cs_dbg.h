#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
void dbg_pdata_i( char *s, int *dp, int size ) ;
void dbg_pdata_c( char *s, char *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( char *s, int *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
void dbg_p_d_data_i_mn_skip ( char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_i_mn ( char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_i_mn_v2 ( char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

// tsting ...
void dbg_p_d_data_ll ( char *s, long long *dp, int size ) ;

#endif 
