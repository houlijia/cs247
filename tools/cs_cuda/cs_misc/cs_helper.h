#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void clear_device_mem_i ( int *d_a, int n ) ;
void set_device_mem_i ( int *d_a, int n, int val ) ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;
void h_expand_c_to_i ( char *din, int *dout, int n ) ;	

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_i ( int *dp, int *hp, int size ) ;

#endif 
