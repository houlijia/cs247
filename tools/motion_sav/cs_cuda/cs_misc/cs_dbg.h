#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
int dbg_copy_d_data ( char *dtp, char *dfp, int size ) ;
void dbg_pdata_i( const char *s, const int *dp, int size ) ;
void dbg_pdata_f( const char *s, const float *dp, int size ) ;
void dbg_pdata_c( const char *s, const char *dp, int size ) ;
void dbg_pdata_d( const char *s, const float *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( const char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( const char *s, int *dp, int size ) ;
void dbg_p_d_data_f ( const char *s, float *dp, int size ) ;
void dbg_p_d_data_d ( const char *s, float *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
float *dbg_d_malloc_f ( int size ) ;

void dbg_p_d_data_f_mn_skip ( const char *s, float *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;
void dbg_p_d_data_i_mn_skip ( const char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint, int printrow ) ;
void dbg_p_d_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( const char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

// print the motion detection T/V/H/Va format
void dbg_p_data_md_f_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_f_mn_v2 ( const char *s, float *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_d_data_i_mn_v2 ( const char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( const char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

void dbg_p_d_data_i_cube ( const char *s, int *devp, int vx, int hy, int tz ) ;
void dbg_p_d_data_i_cube ( const char *s, float *devp, int vx, int hy, int tz ) ;
void dbg_p_data_i_cube ( const char *s, float *dp, int vx, int hy, int tz ) ;

void dbg_pr_first_last ( char *s, float *d_p, int len, int pr_size ) ;
void dbg_pr_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, char *h_p, int len, int pr_size ) ;

// tsting ...
void dbg_p_d_data_ll ( const char *s, long long *dp, int size ) ;

// for random
int dbg_ck_unique ( char *s, int *dp, int size ) ;
int dbg_perm_ck ( int callid, int callid2, int zero_in_zero, int *d_bp, int size ) ;

// TVH
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print, int do_skip ) ;
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print,
	int do_skip, int start ) ;

#endif 
