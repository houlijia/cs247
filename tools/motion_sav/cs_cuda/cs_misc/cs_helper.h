#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void set_device_mem_i ( int *d_a, int n, int val ) ;
void clear_device_mem_i ( int *d_a, int n ) ;

void set_device_mem ( float *d_a, int n, float val ) ;
void clear_device_mem ( float *d_a, int n )	 ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;

void h_expand_c_to_i ( char *din, int *dout, int n ) ;	
void h_expand_c_to_i ( unsigned char *din, int *dout, int n ) ;	
template<typename F, typename T> void h_cast_T_to_T ( F *din, T *dout, int n ) ;

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;
int h_set_cube_config( struct cube *d_a, struct cube *hp ) ;
void h_p_d_cube_config ( const char *, struct cube *dp ) ;
void h_p_h_cube_config ( const char *, struct cube *dp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int put_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes
int get_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes

void h_do_int_to_float ( int *intp, float *fp, int tbl_size ) ;

#endif 
