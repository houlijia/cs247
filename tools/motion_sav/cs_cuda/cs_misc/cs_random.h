#ifndef __CS_RANDOM_H__
#define __CS_RANDOM_H__

int h_set_random_table ( RndC_uint32 seed, RndC_uint32 *h_in, RndC_uint32 *d_out, int size,
	int slot_one_reserved, int do_init )	 ;

#endif 
