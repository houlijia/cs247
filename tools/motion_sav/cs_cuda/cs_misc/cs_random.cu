#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <arpa/inet.h>
#include <math.h>

#include "cs_dbg.h"
#include "cs_cuda.h"
#include "cs_helper.h"

#include "RndCState.h"
#include "RndC_ifc.h"

#include "cs_random.h"

// #define CUDA_DBG
#define CUDA_DBG1

static RndCState cs_random_state ;

// LDL stop here ... in SensingMatrixWH.makepuermutations
// PL = 1 .. 1 + random(order - 1) 
// PR = random(order) 


// val can be either positive or negative

__global__ void 
d_set_random_table( RndC_uint32 *fp, int size, RndC_uint32 val )
{
	int tid = blockIdx.x*blockDim.x + threadIdx.x;

	while ( tid < size )
	{
		fp[ tid ] += val ;
		tid += CUDA_MAX_THREADS ;
	} 
}

/*
    seed : the seed for random number generation
	h_in: points to the host buffer before download to device, if null, a temp buffer will be
		allocated and deleted when done.
	slot_one_reserved: if set [0] will have 0
*/

int
h_set_random_table ( RndC_uint32 seed, RndC_uint32 *h_in, RndC_uint32 *d_out, int size,
	int slot_one_reserved, int do_init )	
{
	int nThreadsPerBlock = CUDA_MAX_THREADS_P_BLK ;
	int nBlocks ;
	RndC_uint32 *h_in_memp = NULL ;
	int i, need_adjustment = 0 ;

	if ( h_in == NULL )
	{
		h_in_memp = ( RndC_uint32 *)malloc ( sizeof ( RndC_uint32 ) * size ) ;
		if ( h_in_memp == NULL )
			return ( 0 ) ;
	} else
		h_in_memp = h_in ;


	if ( do_init )
		init_RndC( &cs_random_state, seed ) ;

	if ( slot_one_reserved )
	{
		h_in_memp[0] = ( RndC_uint32 )0 ;
		randperm1_RndC( &cs_random_state, size - 1, h_in_memp+1 ) ; // from 1 to ( size - 1 )
	} else
	{
		randperm1_RndC( &cs_random_state, size, h_in_memp ) ; // from 1 to ( size )
			// need to subtract one 
		need_adjustment++ ;
	}

	if (( i = cudaMemcpy( d_out, h_in_memp, sizeof ( RndC_uint32 ) * size,
		cudaMemcpyHostToDevice)) != cudaSuccess )
	{
		printf("%s: download cp fail: %d\n", __func__, i ) ;
		
		if ( h_in == NULL )
			free ( h_in_memp ) ;
		return ( 0 ) ;
	}

	if ( need_adjustment )
	{
		h_block_adj ( size, nThreadsPerBlock, &nBlocks ) ;

		d_set_random_table <<< nBlocks, nThreadsPerBlock >>> ( d_out, size, ( RndC_uint32) -1 ) ;

		cudaThreadSynchronize() ;
	}

	if ( h_in == NULL )
		free ( h_in_memp ) ;

	return ( 1 ) ;
}

