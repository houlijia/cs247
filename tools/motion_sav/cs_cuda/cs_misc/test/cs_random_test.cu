#include <iostream>
using namespace std;

#include <thrust/reduce.h>
#include <thrust/sequence.h>
#include <thrust/host_vector.h>
#include <thrust/device_vector.h>

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <math.h>

#include "RndCState.h"
#include "RndC_ifc.h"

#include "cs_dbg.h"
#include "cs_helper.h"
#include "cs_random.h"

#define CUDA_DBG

// #define K64	(1024 * 64)
#define K64	1024

#define TST_SIZE 12

RndC_uint32 *dp1, Krandom[ K64 ], Krandom1[ K64 ] ;
RndC_uint32 seed = 1000 ;

#ifdef CUDA_OBS

// from matlab

>> s = RandStream('mt19937ar','Seed',1000);
>> RandStream.setGlobalStream(s);
>> randperm(11)

	ans =

	     7     2    11     6     9     8     4     1    10     5     3

#endif 

int
main( int ac, char *av[] )
{
	int k ;

	setbuf( stdout, NULL ) ;
	setbuf( stderr, NULL ) ;

	dbg_init ( 1024 * 1024 ) ;

	if (( k = cudaMalloc( &dp1, sizeof ( RndC_uint32 ) * K64 )) != cudaSuccess )
	{
		printf("%s: cube alloc failed %d \n", __func__, k ) ;
		exit ( 9 ) ;
	}

	// no host buf

	if (!( k = h_set_random_table( seed, ( RndC_uint32 * )NULL, ( RndC_uint32 * )dp1, TST_SIZE, 1, 1 )))
	{
		fprintf( stderr, "1 failed\n") ;
		exit( 1 ) ;
	}

	dbg_p_d_data_i("after one", ( int * )dp1, TST_SIZE ) ;  
	dbg_ck_unique ("after one", ( int * )dp1, TST_SIZE ) ;

	if (!( k = h_set_random_table( seed, ( RndC_uint32 * )NULL, ( RndC_uint32 * )dp1, TST_SIZE, 0, 1 )))
	{
		fprintf( stderr, "2 failed\n") ;
		exit( 2 ) ;
	}

	dbg_p_d_data_i("after two", ( int * )dp1, TST_SIZE ) ;  
	dbg_ck_unique ("after two", ( int * )dp1, TST_SIZE ) ;

	// with host buf

	if (!( k = h_set_random_table( seed, Krandom, ( RndC_uint32 * )dp1, TST_SIZE, 1, 1 )))
	{
		fprintf( stderr, "3 failed\n") ;
		exit( 1 ) ;
	}

	dbg_pdata_i("after 3 host", ( int *)Krandom, TST_SIZE ) ;
	dbg_p_d_data_i("after three", ( int * )dp1, TST_SIZE ) ;  
	dbg_ck_unique ("after three", ( int * )dp1, TST_SIZE ) ;

	if (!( k = h_set_random_table( seed, Krandom, ( RndC_uint32 * )dp1, TST_SIZE, 0, 1 )))
	{
		fprintf( stderr, "4 failed\n") ;
		exit( 2 ) ;
	}

	dbg_pdata_i("after 4 host", ( int *)Krandom, TST_SIZE ) ;
	dbg_p_d_data_i("after four", ( int * )dp1, TST_SIZE ) ;  
	dbg_ck_unique ("after four", ( int * )dp1, TST_SIZE ) ;

	printf("only init once ......................................................\n") ;

	// no host buf

	if (!( k = h_set_random_table( seed, ( RndC_uint32 * )NULL, ( RndC_uint32 * )dp1, TST_SIZE, 1, 1 )))
	{
		fprintf( stderr, "1 failed\n") ;
		exit( 1 ) ;
	}

	dbg_p_d_data_i("after one", ( int * )dp1, TST_SIZE ) ;  
	dbg_ck_unique ("after one", ( int * )dp1, TST_SIZE ) ;

	if (!( k = h_set_random_table( seed, ( RndC_uint32 * )NULL, ( RndC_uint32 * )dp1, TST_SIZE, 0, 0 )))
	{
		fprintf( stderr, "2 failed\n") ;
		exit( 2 ) ;
	}

	dbg_p_d_data_i("after two", ( int * )dp1, TST_SIZE ) ;  
	dbg_ck_unique ("after two", ( int * )dp1, TST_SIZE ) ;

	// with host buf

	if (!( k = h_set_random_table( seed, Krandom, ( RndC_uint32 * )dp1, TST_SIZE, 1, 1 )))
	{
		fprintf( stderr, "3 failed\n") ;
		exit( 1 ) ;
	}

	dbg_pdata_i("after 3 host", ( int *)Krandom, TST_SIZE ) ;
	dbg_p_d_data_i("after three", ( int * )dp1, TST_SIZE ) ;  
	dbg_ck_unique ("after three", ( int * )dp1, TST_SIZE ) ;

	if (!( k = h_set_random_table( seed, Krandom, ( RndC_uint32 * )dp1, TST_SIZE, 0, 0 )))
	{
		fprintf( stderr, "4 failed\n") ;
		exit( 2 ) ;
	}

	dbg_pdata_i("after 4 host", ( int *)Krandom, TST_SIZE ) ;
	dbg_p_d_data_i("after four", ( int * )dp1, TST_SIZE ) ;  
	dbg_ck_unique ("after four", ( int * )dp1, TST_SIZE ) ;
}
