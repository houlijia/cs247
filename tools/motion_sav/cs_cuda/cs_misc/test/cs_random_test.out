dbg_p_d_data_i: after one size 48 dp 0x1300200000
dbg_pdata_i: after one
0 -- 00000000 0
1 -- 00000007 7
2 -- 00000002 2
3 -- 0000000b 11
4 -- 00000006 6
5 -- 00000009 9
6 -- 00000008 8
7 -- 00000004 4
8 -- 00000001 1
9 -- 0000000a 10
10 -- 00000005 5
11 -- 00000003 3
dbg_ck_unique : after one good
dbg_p_d_data_i: after two size 48 dp 0x1300200000
dbg_pdata_i: after two
0 -- 00000006 6
1 -- 00000001 1
2 -- 0000000a 10
3 -- 00000005 5
4 -- 00000008 8
5 -- 00000007 7
6 -- 00000003 3
7 -- 00000000 0
8 -- 0000000b 11
9 -- 00000009 9
10 -- 00000004 4
11 -- 00000002 2
dbg_ck_unique : after two good
dbg_pdata_i: after 3 host
0 -- 00000000 0
1 -- 00000007 7
2 -- 00000002 2
3 -- 0000000b 11
4 -- 00000006 6
5 -- 00000009 9
6 -- 00000008 8
7 -- 00000004 4
8 -- 00000001 1
9 -- 0000000a 10
10 -- 00000005 5
11 -- 00000003 3
dbg_p_d_data_i: after three size 48 dp 0x1300200000
dbg_pdata_i: after three
0 -- 00000000 0
1 -- 00000007 7
2 -- 00000002 2
3 -- 0000000b 11
4 -- 00000006 6
5 -- 00000009 9
6 -- 00000008 8
7 -- 00000004 4
8 -- 00000001 1
9 -- 0000000a 10
10 -- 00000005 5
11 -- 00000003 3
dbg_ck_unique : after three good
dbg_pdata_i: after 4 host
0 -- 00000007 7
1 -- 00000002 2
2 -- 0000000b 11
3 -- 00000006 6
4 -- 00000009 9
5 -- 00000008 8
6 -- 00000004 4
7 -- 00000001 1
8 -- 0000000c 12
9 -- 0000000a 10
10 -- 00000005 5
11 -- 00000003 3
dbg_p_d_data_i: after four size 48 dp 0x1300200000
dbg_pdata_i: after four
0 -- 00000006 6
1 -- 00000001 1
2 -- 0000000a 10
3 -- 00000005 5
4 -- 00000008 8
5 -- 00000007 7
6 -- 00000003 3
7 -- 00000000 0
8 -- 0000000b 11
9 -- 00000009 9
10 -- 00000004 4
11 -- 00000002 2
dbg_ck_unique : after four good
only init once ......................................................
dbg_p_d_data_i: after one size 48 dp 0x1300200000
dbg_pdata_i: after one
0 -- 00000000 0
1 -- 00000007 7
2 -- 00000002 2
3 -- 0000000b 11
4 -- 00000006 6
5 -- 00000009 9
6 -- 00000008 8
7 -- 00000004 4
8 -- 00000001 1
9 -- 0000000a 10
10 -- 00000005 5
11 -- 00000003 3
dbg_ck_unique : after one good
dbg_p_d_data_i: after two size 48 dp 0x1300200000
dbg_pdata_i: after two
0 -- 00000009 9
1 -- 00000004 4
2 -- 00000002 2
3 -- 0000000b 11
4 -- 00000001 1
5 -- 00000008 8
6 -- 00000000 0
7 -- 00000003 3
8 -- 00000005 5
9 -- 00000007 7
10 -- 00000006 6
11 -- 0000000a 10
dbg_ck_unique : after two good
dbg_pdata_i: after 3 host
0 -- 00000000 0
1 -- 00000007 7
2 -- 00000002 2
3 -- 0000000b 11
4 -- 00000006 6
5 -- 00000009 9
6 -- 00000008 8
7 -- 00000004 4
8 -- 00000001 1
9 -- 0000000a 10
10 -- 00000005 5
11 -- 00000003 3
dbg_p_d_data_i: after three size 48 dp 0x1300200000
dbg_pdata_i: after three
0 -- 00000000 0
1 -- 00000007 7
2 -- 00000002 2
3 -- 0000000b 11
4 -- 00000006 6
5 -- 00000009 9
6 -- 00000008 8
7 -- 00000004 4
8 -- 00000001 1
9 -- 0000000a 10
10 -- 00000005 5
11 -- 00000003 3
dbg_ck_unique : after three good
dbg_pdata_i: after 4 host
0 -- 0000000a 10
1 -- 00000005 5
2 -- 00000003 3
3 -- 0000000c 12
4 -- 00000002 2
5 -- 00000009 9
6 -- 00000001 1
7 -- 00000004 4
8 -- 00000006 6
9 -- 00000008 8
10 -- 00000007 7
11 -- 0000000b 11
dbg_p_d_data_i: after four size 48 dp 0x1300200000
dbg_pdata_i: after four
0 -- 00000009 9
1 -- 00000004 4
2 -- 00000002 2
3 -- 0000000b 11
4 -- 00000001 1
5 -- 00000008 8
6 -- 00000000 0
7 -- 00000003 3
8 -- 00000005 5
9 -- 00000007 7
10 -- 00000006 6
11 -- 0000000a 10
dbg_ck_unique : after four good
